metadata-ingestion/src/datahub/ingestion/source/snaplogic/snaplogic.py

```diff
@@ -0,0 +1,349 @@
+            )
+
+        def field_urn(d, f):
+            return make_schema_field_urn(dataset_urn(d), f["name"])
+
+        # Emit lineage
+        yield MetadataChangeProposalWrapper(
+            entityUrn=job_urn,
+            aspect=DataJobInputOutputClass(
+                inputDatasets=[dataset_urn(d) for d in input_datasets],
+                outputDatasets=[dataset_urn(d) for d in output_datasets],
+                inputDatasetFields=[
+                    field_urn(d, f) for d in input_datasets for f in d.fields
+                ],
+                outputDatasetFields=[
+                    field_urn(d, f) for d in output_datasets for f in d.fields
+                ],
+                fineGrainedLineages=[
+                    FineGrainedLineageClass(
+                        upstreamType=FineGrainedLineageDownstreamTypeClass.FIELD_SET,
+                        upstreams=[
+                            make_schema_field_urn(
+                                make_dataset_urn_with_platform_instance(
+                                    cl.input_dataset.platform,
+                                    cl.input_dataset.name,
+                                    cl.input_dataset.platform_instance,
+                                    cl.input_dataset.env,
+                                ),
+                                cl.input_field,
+                            )
+                        ],
+                        downstreamType=FineGrainedLineageDownstreamTypeClass.FIELD_SET,
+                        downstreams=[
+                            make_schema_field_urn(
+                                make_dataset_urn_with_platform_instance(
+                                    cl.output_dataset.platform,
+                                    cl.output_dataset.name,
+                                    cl.output_dataset.platform_instance,
+                                    cl.output_dataset.env,
+                                ),
+                                cl.output_field,
+                            )
+                        ],
+                    )
+                    for cl in columns_mapping
+                ],
+            ),
+        ).as_workunit()
+
+    def create_dataset_mcp(
+        self,
+        dataset_name: str,
+        dataset_display_name: str,
+        fields: list[dict],
+        platform: str = "snaplogic",
+        env: str = "PROD",
+        platform_instance: Optional[str] = None,
+    ) -> Iterable[MetadataWorkUnit]:
+        dataset_urn = make_dataset_urn_with_platform_instance(
+            platform=platform,
+            name=dataset_name,
+            env=env,
+            platform_instance=platform_instance,
+        )
+        # If we have datasets already in the datahub we don't need to create it.
+        # We only create datasets for other platforms if they don't exist in the datahub.
+        if platform != "snaplogic" and self.graph and self.graph.exists(dataset_urn):
```

Comment:
This can potentially be expensive if you have tons of datasets. That would mean one graph query per dataset, which won't scale.

I would implement this behind a feature flag and not automatically create/materialize datasets.
In other sources, we just emit lineage edges for datasets and expect users to run platform-specific  sources to materialize datasets.